test/evaluate/ISO-Fortran-binding.cc

```diff
@@ -210,14 +210,459 @@ int main() {
+
+  resJ = result->dim[1].lower_bound;
+  for (CFI_index_t k{lb[2]}; k <= ub[2]; k += strides[2]) {
+    for (CFI_index_t j{lb[1]}; j <= ub[1]; j += 1) {
+      CFI_index_t resI{result->dim[1].lower_bound + result->dim[0].extent - 1};
+      for (CFI_index_t i{2}; i <= 4; ++i) {
+        // check A(i,j,k) == B(resI, resJ) == array[k-1][j-1][i-1]
+        const CFI_index_t resSubcripts[] = {resI, resJ};
+        const CFI_index_t srcSubcripts[] = {i, j, k};
+        MATCH(true,
+            CFI_address(source, srcSubcripts) ==
+                CFI_address(result, resSubcripts));
+        MATCH(true,
+            CFI_address(source, srcSubcripts) ==
+                &array[k - lbs2][j - lbs1][i - lbs0]);
+        --resI;
+      }
+    }
+    ++resJ;
+  }
+}
+
+static void run_CFI_select_part_tests() {
+  constexpr std::size_t name_len = 5;
+  typedef struct {
+    double distance;
+    int stars;
+    char name[name_len];
+  } Galaxy;
+
+  const CFI_rank_t rank{2};
+  constexpr CFI_index_t universeSize[] = {2, 3};
+  Galaxy universe[universeSize[1]][universeSize[0]];
+
+  for (int j{0}; j < universeSize[1]; ++j) {
+    for (int i{0}; i < universeSize[0]; ++i) {
+      universe[j][i].distance = i + j * 32;
+      universe[j][i].stars = i * 2 + j * 64;
+      universe[j][i].name[2] = static_cast<char>(i);
+      universe[j][i].name[3] = static_cast<char>(j);
+    }
+  }
+
+  CFI_CDESC_T(rank) resStorage, srcStorage;
+  CFI_cdesc_t *result = reinterpret_cast<CFI_cdesc_t *>(&resStorage);
+  CFI_cdesc_t *source = reinterpret_cast<CFI_cdesc_t *>(&srcStorage);
+
+  bool testPreConditions{true};
+  int retCode = CFI_establish(result, nullptr, CFI_attribute_other,
+      CFI_type_int, sizeof(int), rank, nullptr);
+  testPreConditions &= (retCode == CFI_SUCCESS);
+  retCode = CFI_establish(source, &universe, CFI_attribute_other,
+      CFI_type_struct, sizeof(Galaxy), rank, universeSize);
+  testPreConditions &= (retCode == CFI_SUCCESS);
+  if (!testPreConditions) {
+    MATCH(true, testPreConditions);
+    return;
+  }
+
+  std::size_t displacement{offsetof(Galaxy, stars)};
+  std::size_t elem_len{0};  // ignored
+  retCode = CFI_select_part(result, source, displacement, elem_len);
+  MATCH(CFI_SUCCESS, retCode);
+
+  bool baseAddrShiftedOk =
+      (reinterpret_cast<char *>(source->base_addr) + displacement ==
+          result->base_addr);
+  MATCH(true, baseAddrShiftedOk);
+  if (!baseAddrShiftedOk) {
+    return;
+  }
+
+  MATCH(sizeof(int), result->elem_len);
+  for (CFI_index_t j{0}; j < universeSize[1]; ++j) {
+    for (CFI_index_t i{0}; i < universeSize[0]; ++i) {
+      CFI_index_t subscripts[]{
+          result->dim[0].lower_bound + i, result->dim[1].lower_bound + j};
+      MATCH(i * 2 + j * 64,
+          *reinterpret_cast<int *>(CFI_address(result, subscripts)));
+    }
+  }
+
+  // Test for Fortran character type
+  retCode = CFI_establish(
+      result, nullptr, CFI_attribute_other, CFI_type_char, 2, rank, nullptr);
```

Comment:
Ok, I may have missed something here. Let me explain you why I believe CFI_type_char is Fortran characters for the following reasons:

- Table 18.4 (in section 18.5.4):
The table is split in four parts: The first regarding integers contains CFI_type_signed_char for C `signed char`. Then there is a part on real types and one on complex types. Last, there is a part with what I understand are non numerical types. This parts defines CFI_type_char for C `char`. These lead me to understand that CFI_type_char refers to Fortran CHARACTER when CFI_type_signed_char is INTEGER*1.
- CFI_type_cptr is associated in this table to void* that I understand as the C pointer types rather than Fortran CHARACTER. I understood it as linked to the C_PTR type described in sections 18.3.2, though I am not sure they directly relate.
- In table 18.2 (in section 18.3.1):
Fortran CHARACTER is associated to C_CHAR from ISO_C_BINDING module and to the C type `char`. Note 2 of 18.5.3 links C_CHAR, and hence `char` to Fortran CHARACTER type in the context of descriptors.
In table 18.2, the standard also refers to C `unsigned char` and `signed char` for the numerical types INTEGER*1.

Again, I may just have skipped an important paragraph. I will give it a deeper look then.

Reply:
In 18.3.1 paragraph 1 3rd sentence, the standard mentions that the CHARACTER to C_CHAR/char mapping in table 18.2 applies if and only if the LEN type parameter is 1. Note that CHARACTER values with LEN > 1 cannot be represented in C's char type.